cmd/pipeline-controller/reconciler.go

```diff
@@ -0,0 +1,220 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/kube"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	retention = 24 * time.Hour
+)
+
+type minimalGhClient interface {
+	GetPullRequest(org, repo string, number int) (*github.PullRequest, error)
+	CreateComment(org, repo string, number int, comment string) error
+}
+
+type pullRequest struct {
+	closed    bool
+	checkTime time.Time
+}
+
+type closedPRsCache struct {
+	prs       map[string]pullRequest
+	m         sync.Mutex
+	ghc       minimalGhClient
+	clearTime time.Time
+}
+
+func composePRIdentifier(refs *v1.Refs) string {
+	return fmt.Sprintf("%s/%s/%d", refs.Org, refs.Repo, refs.Pulls[0].Number)
+}
+
+func (c *closedPRsCache) isPRClosed(refs *v1.Refs) (bool, error) {
+	id := composePRIdentifier(refs)
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.clearCache()
+	pr, ok := c.prs[id]
+	if ok && (time.Since(pr.checkTime) < 5*time.Minute || pr.closed) {
+		return pr.closed, nil
+	}
+	ghPr, err := c.ghc.GetPullRequest(refs.Org, refs.Repo, refs.Pulls[0].Number)
+	if err != nil {
+		return false, fmt.Errorf("error getting pull request: %w", err)
+	}
+	c.prs[id] = pullRequest{closed: ghPr.State != github.PullRequestStateOpen, checkTime: time.Now()}
+	return ghPr.State != github.PullRequestStateOpen, nil
+}
+
+func (c *closedPRsCache) clearCache() {
+	if time.Since(c.clearTime) < retention {
+		return
+	}
+	for k, v := range c.prs {
+		if time.Since(v.checkTime) >= retention {
+			delete(c.prs, k)
+		}
+	}
+	c.clearTime = time.Now()
+}
+
+func composeKey(refs *v1.Refs) string {
+	return fmt.Sprintf("%s/%s/%d/%s/%s", refs.Org, refs.Repo, refs.Pulls[0].Number, refs.BaseRef, refs.Pulls[0].SHA)
+}
+
+type reconciler struct {
+	pjclientset        ctrlruntimeclient.Client
+	lister             ctrlruntimeclient.Reader
+	configDataProvider *ConfigDataProvider
+	ghc                minimalGhClient
+	closedPRsCache     closedPRsCache
+	ids                sync.Map
```

Comment:
Can you please add a comment describing what's the purpose of this field `ids`?

Reply:
I can change a name, as I am not a big fan of comments :)